crates/flux-middle/src/fhir.rs

```diff
@@ -558,10 +561,27 @@ pub enum Lifetime {
 
 #[derive(Clone, Copy)]
 pub struct ArrayLen {
-    pub val: usize,
+    pub kind: ArrayLenKind,
     pub span: Span,
 }
 
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub enum ArrayLenKind {
+    /// The length of the array is a constant
+    Lit(usize),
+    /// The length of the array is a type parameter
+    ParamConst(ParamConst),
```

Comment:
Hmm in `conv.rs` we need to convert 

```rust
   fhir::TyKind::Array(ty, len) => {
                Ok(rty::Ty::array(
                    self.conv_ty(env, ty)?,
                    rty::Const::from_array_len(self.genv.tcx(), len.kind),
                ))
            }
``` 

This in turn requires

```rust
 pub fn from_array_len(tcx: TyCtxt, len: ArrayLenKind) -> Const {
        let kind = match len {
            ArrayLenKind::Lit(len) => {
                ConstKind::Value(ScalarInt::try_from_target_usize(len as u128, tcx).unwrap())
            }

            ArrayLenKind::ParamConst(param_const) => ConstKind::Param(param_const),
        };
        Const { kind, ty: Ty::mk_uint(UintTy::Usize) }
    }
```

Specifically the

```rust
            ArrayLenKind::ParamConst(param_const) => ConstKind::Param(param_const),
``` 

indicates when we're conv-ing from the `fhir::ParamConst(ParamConst)` the corresponding translated thing needs to be a `ParamConst`? Does it make sense to do the gymnastics to turn it back into `DefID` there? (and how even -- we'd need to somehow tweak the `conv` to track the `generics_of` ? )

Reply:
Ah I suppose one can use the `OwnerId` stashed in the `WfckResults`?

Reply:
No, it's simpler than that. See what we do for `ParamTy` here https://github.com/flux-rs/flux/blob/main/crates/flux-middle/src/global_env.rs#L325.

We can do the same for `ParamConst`

```rust
    pub fn def_id_to_param_const(&self, def_id: LocalDefId) -> rty::ParamConst {
        rty::ParamConst {
            index: self.def_id_to_param_index(def_id),
            name: self.tcx().hir().ty_param_name(def_id),
        }
    }
 ```

Reply:
I could be missing something as severely jet lagged, but the issue is we need to go from 

`ParamConst --> DefId`

(Not the other way around...)

Reply:
In particular, in the 


```
            ArrayLenKind::ParamConst(param_const) => ConstKind::Param(param_const)
```

You want to make the RHS be a `def_id` so we need to convert the `param_const` on the LHS to a `def_id` on the RHS.

Reply:
No, the opposite, the lhs should be a `def_id`. 

Also,  `from_array_len` should not be a function in `Const` because `rustc::ty` should not depend on `fhir`.

Reply:
moved the `Const` stuff out too.